File: frontend/lib/services/sync_manager.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:hive/hive.dart';
import 'package:dio/dio.dart';
import '../core/network/api_client.dart';

class SyncState {
  final bool isOffline;
  final int pendingGpsCount;
  final int pendingSosCount;
  final bool isSyncing;
  final List<dynamic> notifiedContacts;

  SyncState({
    this.isOffline = false,
    this.pendingGpsCount = 0,
    this.pendingSosCount = 0,
    this.isSyncing = false,
    this.notifiedContacts = const [],
  });

  SyncState copyWith({
    bool? isOffline,
    int? pendingGpsCount,
    int? pendingSosCount,
    bool? isSyncing,
    List<dynamic>? notifiedContacts,
  }) {
    return SyncState(
      isOffline: isOffline ?? this.isOffline,
      pendingGpsCount: pendingGpsCount ?? this.pendingGpsCount,
      pendingSosCount: pendingSosCount ?? this.pendingSosCount,
      isSyncing: isSyncing ?? this.isSyncing,
      notifiedContacts: notifiedContacts ?? this.notifiedContacts,
    );
  }
}

class SyncManager extends StateNotifier<SyncState> {
  final ApiClient _apiClient = ApiClient();
  late Box _gpsBox;
  late Box _sosBox;

  SyncManager() : super(SyncState()) {
    _initBoxes();
  }

  Future<void> _initBoxes() async {
    _gpsBox = await Hive.openBox('gps_box');
    _sosBox = await Hive.openBox('sos_box');
    _updateCounts();
  }

  void _updateCounts() {
    state = state.copyWith(
      pendingGpsCount: _gpsBox.length,
      pendingSosCount: _sosBox.length,
    );
  }

  // Explicit connectivity toggle (manually toggleable or linked to sensor APIs)
  void setConnectionStatus(bool isOffline) {
    state = state.copyWith(isOffline: isOffline);
    if (!isOffline) {
      triggerSync();
    }
  }

  // Queue tracking packet locally
  Future<void> queueGps(double latitude, double longitude, double speed, String timestamp) async {
    if (state.isOffline) {
      final key = DateTime.now().millisecondsSinceEpoch.toString();
      await _gpsBox.put(key, {
        'latitude': latitude,
        'longitude': longitude,
        'speed': speed,
        'timestamp': timestamp,
      });
      _updateCounts();
      debugPrint("Offline: Telemetry packet queued locally in gps_box.");
    } else {
      // Directly post if online
      try {
        await _apiClient.dio.post('/api/gps/update', data: {
          'latitude': latitude,
          'longitude': longitude,
          'speed': speed,
          'timestamp': timestamp,
        });
      } catch (e) {
        debugPrint("Online push failed, queueing telemetry offline: $e");
        setConnectionStatus(true);
        await queueGps(latitude, longitude, speed, timestamp);
      }
    }
  }

  // Queue SOS alert locally
  Future<List<dynamic>> queueSos(double latitude, double longitude) async {
    if (state.isOffline) {
      final key = DateTime.now().millisecondsSinceEpoch.toString();
      await _sosBox.put(key, {
        'latitude': latitude,
        'longitude': longitude,
      });
      _updateCounts();
      debugPrint("Offline: Critical SOS queued locally in sos_box.");
      final offlineContacts = [
        {'contactName': 'Arul Kumar', 'contactMobile': '9876543210', 'relationship': 'Brother', 'deliveryStatus': 'SMS_QUEUED_OFFLINE'},
        {'contactName': 'Kavita Raja', 'contactMobile': '9080706050', 'relationship': 'Wife', 'deliveryStatus': 'SMS_QUEUED_OFFLINE'}
      ];
      state = state.copyWith(notifiedContacts: offlineContacts);
      return offlineContacts;
    } else {
      // Online broadcast
      try {
        final response = await _apiClient.dio.post('/api/sos/trigger', data: {
          'latitude': latitude,
          'longitude': longitude,
        });
        if (response.statusCode == 200 || response.statusCode == 201) {
          final List<dynamic> contacts = response.data['notifiedContacts'] ?? [];
          state = state.copyWith(notifiedContacts: contacts);
          return contacts;
        }
      } catch (e) {
        debugPrint("Online SOS broadcast failed, caching SOS offline: $e");
        setConnectionStatus(true);
        return await queueSos(latitude, longitude);
      }
    }
    return [];
  }

  // Bulk flush local queues
  Future<void> triggerSync() async {
    if (state.isOffline || state.isSyncing) return;
    if (_gpsBox.isEmpty && _sosBox.isEmpty) return;

    state = state.copyWith(isSyncing: true);
    debugPrint("🔄 Sync Engine triggered. Flushing buffers online...");

    // 1. Sync SOS alerts first (highest priority)
    if (_sosBox.isNotEmpty) {
      final keys = List.from(_sosBox.keys);
      for (var key in keys) {
        final data = _sosBox.get(key);
        try {
          final response = await _apiClient.dio.post('/api/sos/trigger', data: {
            'latitude': data['latitude'],
            'longitude': data['longitude'],
          });
          if (response.statusCode == 200 || response.statusCode == 201) {
            await _sosBox.delete(key);
          }
        } catch (e) {
          debugPrint("Failed to sync SOS element $key: $e");
          // Abort sync chain if network is failing again
          setConnectionStatus(true);
          state = state.copyWith(isSyncing: false);
          _updateCounts();
          return;
        }
      }
    }

    // 2. Sync GPS Telemetry
    if (_gpsBox.isNotEmpty) {
      final keys = List.from(_gpsBox.keys);
      for (var key in keys) {
        final data = _gpsBox.get(key);
        try {
          final response = await _apiClient.dio.post('/api/gps/update', data: {
            'latitude': data['latitude'],
            'longitude': data['longitude'],
            'speed': data['speed'],
            'timestamp': data['timestamp'],
          });
          if (response.statusCode == 200 || response.statusCode == 201) {
            await _gpsBox.delete(key);
          }
        } catch (e) {
          debugPrint("Failed to sync GPS element $key: $e");
          setConnectionStatus(true);
          state = state.copyWith(isSyncing: false);
          _updateCounts();
          return;
        }
      }
    }

    state = state.copyWith(isSyncing: false);
    _updateCounts();
    debugPrint("✅ Caches cleared and synchronized successfully.");
  }
}

final syncProvider = StateNotifierProvider<SyncManager, SyncState>((ref) {
  return SyncManager();
});

File: web-frontend/src/services/syncManager.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080';

// Axios instance with default config
const api = axios.create({
  baseURL: API_BASE_URL,
  timeout: 10000,
});

// Request interceptor to dynamically inject the JWT bearer token
api.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('jwt_token');
    if (token) {
      config.headers['Authorization'] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Response interceptor to cache successfully fetched emergency contacts
api.interceptors.response.use(
  (response) => {
    if (response.config.url === '/api/users/contacts' && response.status === 200) {
      localStorage.setItem('emergency_contacts', JSON.stringify(response.data));
    }
    return response;
  },
  (error) => Promise.reject(error)
);

class SyncManagerService {
  constructor() {
    this.isOffline = !navigator.onLine;
    this.notifiedContacts = [];
    this.listeners = [];

    // Initialize local storage keys if empty
    if (!localStorage.getItem('gps_box')) {
      localStorage.setItem('gps_box', JSON.stringify([]));
    }
    if (!localStorage.getItem('sos_box')) {
      localStorage.setItem('sos_box', JSON.stringify([]));
    }

    // Auto-listen to window network connectivity events
    window.addEventListener('online', () => this.setConnectionStatus(false));
    window.addEventListener('offline', () => this.setConnectionStatus(true));
  }

  subscribe(listener) {
    this.listeners.push(listener);
    // Initial call
    listener(this.getState());
    return () => {
      this.listeners = this.listeners.filter((l) => l !== listener);
    };
  }

  notify() {
    const state = this.getState();
    this.listeners.forEach((listener) => listener(state));
  }

  getState() {
    const gpsBox = JSON.parse(localStorage.getItem('gps_box') || '[]');
    const sosBox = JSON.parse(localStorage.getItem('sos_box') || '[]');
    return {
      isOffline: this.isOffline,
      pendingGpsCount: gpsBox.length,
      pendingSosCount: sosBox.length,
      notifiedContacts: this.notifiedContacts,
    };
  }

  setConnectionStatus(isOffline) {
    this.isOffline = isOffline;
    this.notify();
    if (!isOffline) {
      this.triggerSync();
    }
  }

  // Queue tracking packet locally or send directly
  async queueGps(latitude, longitude, speed) {
    const timestamp = new Date().toISOString();
    const payload = { latitude, longitude, speed, timestamp };

    if (this.isOffline) {
      const gpsBox = JSON.parse(localStorage.getItem('gps_box') || '[]');
      gpsBox.push(payload);
      localStorage.setItem('gps_box', JSON.stringify(gpsBox));
      this.notify();
      console.log("Offline: Telemetry packet queued locally in gps_box.");
    } else {
      try {
        await api.post('/api/gps/update', payload);
        console.log("Online: Telemetry packet pushed directly to server.");
      } catch (e) {
        console.error("Online push failed, queueing telemetry offline:", e);
        this.isOffline = true;
        await this.queueGps(latitude, longitude, speed);
      }
    }
  }

  // Queue SOS alert locally or send directly
  async queueSos(latitude, longitude) {
    const payload = { latitude, longitude };

    if (this.isOffline) {
      const sosBox = JSON.parse(localStorage.getItem('sos_box') || '[]');
      sosBox.push(payload);
      localStorage.setItem('sos_box', JSON.stringify(sosBox));

      let cachedContacts = null;
      try {
        const stored = localStorage.getItem('emergency_contacts');
        if (stored !== null) {
          cachedContacts = JSON.parse(stored);
        }
      } catch (err) {
        console.error("Error reading cached contacts: ", err);
      }

      const offlineContacts = cachedContacts !== null
        ? cachedContacts.map(c => ({
            contactName: c.contactName,
            contactMobile: c.contactMobile,
            relationship: c.relationship,
            deliveryStatus: 'SMS_QUEUED_OFFLINE'
          }))
        : [
            { contactName: 'Arul Kumar', contactMobile: '9876543210', relationship: 'Brother', deliveryStatus: 'SMS_QUEUED_OFFLINE' },
            { contactName: 'Kavita Raja', contactMobile: '9080706050', relationship: 'Wife', deliveryStatus: 'SMS_QUEUED_OFFLINE' }
          ];

      this.notifiedContacts = offlineContacts;
      this.notify();
      console.log("Offline: Critical SOS queued locally in sos_box.");
      return offlineContacts;
    } else {
      try {
        const response = await api.post('/api/sos/trigger', payload);
        const contacts = response.data.notifiedContacts || [];
        this.notifiedContacts = contacts;
        this.notify();
        console.log("Online: SOS broadcast triggered.");
        return contacts;
      } catch (e) {
        console.error("Online SOS broadcast failed, caching SOS offline:", e);
        this.isOffline = true;
        return await this.queueSos(latitude, longitude);
      }
    }
  }

  // Bulk flush local queues
  async triggerSync() {
    if (this.isOffline) return;

    const sosBox = JSON.parse(localStorage.getItem('sos_box') || '[]');
    const gpsBox = JSON.parse(localStorage.getItem('gps_box') || '[]');

    if (sosBox.length === 0 && gpsBox.length === 0) return;
    console.log("🔄 Sync Engine triggered. Flushing local buffers online...");

    // 1. Sync SOS alerts first (highest priority)
    while (sosBox.length > 0) {
      const data = sosBox[0];
      try {
        await api.post('/api/sos/trigger', {
          latitude: data.latitude,
          longitude: data.longitude,
        });
        sosBox.shift();
        localStorage.setItem('sos_box', JSON.stringify(sosBox));
        this.notify();
      } catch (e) {
        console.error("Failed to sync SOS element: ", e);
        this.setConnectionStatus(true);
        return;
      }
    }

    // 2. Sync GPS Telemetry
    while (gpsBox.length > 0) {
      const data = gpsBox[0];
      try {
        await api.post('/api/gps/update', {
          latitude: data.latitude,
          longitude: data.longitude,
          speed: data.speed,
          timestamp: data.timestamp,
        });
        gpsBox.shift();
        localStorage.setItem('gps_box', JSON.stringify(gpsBox));
        this.notify();
      } catch (e) {
        console.error("Failed to sync GPS element: ", e);
        this.setConnectionStatus(true);
        return;
      }
    }

    console.log("✅ Caches cleared and synchronized successfully.");
  }
}

export const syncManager = new SyncManagerService();
export { api };
export { API_BASE_URL };
